[Unity][ONNX] Apply `R.nn.attention` in Onnx Attention OP

cyx-6 opened the pull request:
This PR refactors the frontend impl of Onnx attention op, to apply the `R.nn.attention`. So we are able to use the cutlass BYOC with Onnx frontend, or other backends for `R.nn.attention` in the future. And this PR adds more support over Onnx attention op inputs and attributes.

hypercubestart commented on python/tvm/relax/block_builder.py:
```diff
@@ -388,25 +388,24 @@ def rx_func(x: Tensor[(n,), "float32"], y: Tensor[((n + 1),), "float32"]) -> Ten
-        # TODO(hypercubestart, ziheng) handle multiple output case
         te_out = func(*new_args, **new_kwargs)
-        assert isinstance(te_out, tvm.te.tensor.Tensor), "only support te tensor as function output"
+        outs = [te_out] if isinstance(te_out, tvm.te.tensor.Tensor) else list(te_out)
```
is it possible for Tensors to be nested in a recursive data structure? such as (A, (B, (C, D, E)))?

YuchenJin replied:
It seems possible: https://discuss.tvm.apache.org/t/graph-plan-memory-doesnt-support-nested-tuples/8278

hypercubestart replied:
I think this post is referring to Relay Tuples. Sorry if I wasn't clear, is it a common use case to use recursive data structure with te compute functions? 

i.e 

```python
def te_compute(A):
   B = te.compute(...)
   C = te.compute(...)
   D = te.compute(...)
   (B, (C, D))
```

I looked into the TVM codebase but couldn't find anything more than a single-level tuple

YuchenJin replied:
Thanks for the clarification! I'm not sure about this. Feel free to bring it up in the open dev meeting. :)

hypercubestart commented on python/tvm/relax/block_builder.py:
```diff
@@ -488,6 +487,7 @@ def emit_func_output(self,
         if len(block.bindings) > 0:
             self._blocks.append(block)
         seqe = rx.SeqExpr(self._blocks, self._func_ret)
+        # TODO: rx function can return Tuple as well, need to infer ret type
```
can we infer the function ret type here?

YuchenJin replied:
I think we can analyze the `checked_type` of the vars in emit_func_output([var0, var1, ...]) , and fill them as the `ret_type` of the function (currently we just set it to rx.DynTensorType(-1)).

hypercubestart replied:
agreed, I think an issue with the current state of Relax is that we don't really enforce imperative type propagation (since using rx.DynTensorType(-1) is always correct). This worked when the only type was Tensors but now when we introduce Tuples, we will need to infer the _checked_type field correctly

hypercubestart commented on src/relax/op/op.cc:
```diff
@@ -69,7 +69,18 @@ Expr MakeCallTIR(Expr shape, Expr func, Tuple args, Optional<Expr> packed_ints)
+    call->checked_type_ = TupleType(types);
+  } else {
+    // TODO: fix checked_type_ inference
+    call->checked_type_ = args->fields[0]->checked_type_;
```
how to infer the dtype for checked_type_ here?

YuchenJin replied:
I think we might need to deduce the dtype of the outputs by analyzing the PrimFunc or the PackedFunc being called.

It's also related to one discussion Ziheng and I had about whether we want to explicitly have the output shape as the argument of `call_tir`. On one hand, it's good info; on the other hand, the same info is available in `call.shape_`. Also, the output dtype is maintained in the `call.checked_type` instead of its arguments. It's about whether we want to make dtype and shape treatment consistent.

hypercubestart replied:
hmm okay, I think we may need to pass the dtype information as an input then so that type information is propagated correctly

YuchenJin replied:
Exactly! The shape and dtype can be user-provided inputs to `call_tir`. 
From the text format perspective, we can have the following syntactic sugar to look into the shape and dtype info in the lhs:
`x: Tensor[shape, dtype] = call_tir(lowlevel_func, [a, b])`

hypercubestart commented:
based on discussion from the meeting, we want to add testcase for https://tvm.apache.org/docs/how_to/work_with_schedules/tuple_inputs.html . It looks like the argmax test case (reducenode with multiple sources) is currently unsupported by create_primfunc.

I think we may want to split the PR into the following:
- [ ] call tir multi output semantics + Tuple support (this one)
- [ ] extend create_primfunc support to cover reduce node with multiple sources (this can then be easily upstreamed to main TVM)
- [ ] syntactical sugar to pass dtype to call tir (Yuchen's comment)
- [ ] changes to script/parser to support Tuple (from discussion)

@YuchenJin @ZihengJiang does this sound good?

YuchenJin commented:
@hypercubestart, it sounds good to me. Good job breaking down the goal into steps!

ZihengJiang commented:
Sounds good to me! @hypercubestart

ZihengJiang commented on tests/python/relax/test_vm.py:
```diff
@@ -607,6 +607,31 @@ def te_func(A):
+    with bb.function("rx_func"):
+        x = nn.Placeholder((n,), dtype="float32", name="x")
+        y = nn.Placeholder((n,), dtype="float32", name="y")
+        tup = relay.Tuple([x, y])
```
import name Tuple into relax so that we don't need to import them both

ZihengJiang commented on tests/python/relax/test_vm.py:
```diff
@@ -607,6 +607,31 @@ def te_func(A):
+        x = nn.Placeholder((n,), dtype="float32", name="x")
+        y = nn.Placeholder((n,), dtype="float32", name="y")
+        tup = relay.Tuple([x, y])
+        getitem = relay.TupleGetItem(tup, 0)
```
use `tup[0]`

ZihengJiang commented on python/tvm/relax/block_builder.py:
```diff
@@ -388,25 +388,24 @@ def rx_func(x: Tensor[(n,), "float32"], y: Tensor[((n + 1),), "float32"]) -> Ten
 
-        call_args = [x.op.value for x in inputs[:-1]]
+        call_args = [x.op.value for x in te_args]
+        output_shape = outs[0].shape if isinstance(te_out, tvm.te.tensor.Tensor) else Tuple([ShapeExpr(x.shape) for x in outs])
```
raise error for unsupported type

hypercubestart replied:
thanks for the review! added the check above

ZihengJiang commented:
LGTM. The PR can be merged once the conflicts can be resolved.

cyx-6 committed:
[Unity][ONNX] Apply `R.nn.attention` in Onnx Attention OP

This PR refactors the frontend impl of Onnx attention op, to apply the `R.nn.attention`. So we are able to use the cutlass BYOC with Onnx frontend, or other backends for `R.nn.attention` in the future.
And this PR adds more support over Onnx attention op inputs and attributes.

cyx-6 committed:
fix lint

cyx-6 committed:
fix lint

jwfromm commented:
Awesome, thanks for this improvement Yaxing!
